salt/orch/removal.sls

```diff
@@ -0,0 +1,394 @@
+
+{#
+ # replacement for k8s minions
+ #}
+{%- if target in minions %}
+  {%- if not replacement %}
+    {# TODO: implement a replacement finder for k8s minions #}
+  {%- endif %}
+
+  {# check if the replacement provided/found is valid #}
+  {# NOTE: maybe the new role has already been assigned in Velum... #}
+  {%- if replacement %}
+    {%- set bootstrapped_minions = salt.saltutil.runner('mine.get', tgt='G@roles:kube-minion and G@bootstrap_complete:true', fun='network.interfaces', tgt_type='compound').keys() %}
+    {%- if replacement in bootstrapped_minions %}
+      {%- do salt.caasp_log.warn('CaaS: warning! replacement for %s, %s, has already been assigned a k8s minion role', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
+      {%- endif %}
+      {%- set replacement = '' %}
+    {%- elif replacement in masters %}
+      {%- do salt.caasp_log.warn('CaaS: will not replace the k8s minion %s: the replacement %s is already a k8s master', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
+      {%- endif %}
+      {%- set replacement = '' %}
+    {%- elif 'kube-master' in replacement_roles %}
+      {%- do salt.caasp_log.warn('CaaS: will not replace the k8s minion %s: the replacement found/provided, %s, is already scheduled for being a new k8s master', target, replacement) %}
+      {%- if replacement_provided %}
+        {%- do salt.caasp_log.abort('CaaS: fatal!! the user provided replacement %s cannot be used', replacement) %}
+      {%- endif %}
+      {%- set replacement = '' %}
+    {%- endif %}
+  {%- endif %}
+
+  {%- if replacement %}
+    {%- do salt.caasp_log.debug('CaaS: setting %s as replacement for the k8s minion %s', replacement, target) %}
+    {%- do replacement_roles.append('kube-minion') %}
+  {%- elif minions|length == 1 %}
+    {%- do salt.caasp_log.warn('CaaS: number of k8s minions will be 0, as no replacement for %s has been found (or provided)', target) %}
+  {%- endif %}
+{%- endif %} {# target in minions #}
+
+{#- other consistency checks... #}
+{%- if replacement %}
+  {#- consistency check: if there is a replacement, it must have some (new) role(s) #}
+  {%- if not replacement_roles %}
+    {%- do salt.caasp_log.abort('CaaS: %s cannot be removed: too few etcd members, and no replacement found', target) %}
+  {%- endif %}
+{%- endif %} {# replacement #}
+
+{##############################
+ # set grains
+ #############################}
+
+assign-removal-grain:
+  salt.function:
+    - tgt: {{ target }}
+    - name: grains.setval
+    - arg:
+      - removal_in_progress
+      - true
+
+{%- if replacement %}
+
+assign-addition-grain:
+  salt.function:
+    - tgt: {{ replacement }}
+    - name: grains.setval
+    - arg:
+      - addition_in_progress
```

Comment:
`addition_in_progress` looks way too generic, could we rename this to something more specific, like `etcd_member_addition_in_progress` or something similar?

Reply:
It is used for any addition, not only etcd members.

Reply:
A replacement refers basically to `run this etcd member on this replacement when it was running in this node that was to be removed`? In the context of this PR, and in general in our setup, the replacement is strictly referred to `etcd` as far as I can see. No?

Reply:
The replacement is the node that replaces the removed one. This replacement can be provided by the user or auto-calculated. It is not only for etcd: you could replace a k8s master (with no etcd running), so the replacement for this master would be marked with `addition_in_progress`... (so far we are only using this flag for etcd, but it coud be used in other roles too)